src/library/scala/math/BigInt.scala

```diff
@@ -221,84 +252,83 @@ final class BigInt private (private val _bigInteger: BigInteger, private val _lo
 
   /** Compares this BigInt with the specified BigInt for equality.
    */
-  def equals (that: BigInt): Boolean = (this eq that) || (
-    if (this.isValidLong) {
-      if (that.isValidLong) this._long == that._long else false
+  def equals (that: BigInt): Boolean =
+    if (this.longEncoding) {
+      if (that.longEncoding) this._long == that._long else false
     } else {
-      if (that.isValidLong) false else this._bigInteger == that._bigInteger
+      if (that.longEncoding) false else this._bigInteger == that._bigInteger
     }
-    )
 
   /** Compares this BigInt with the specified BigInt
    */
   def compare (that: BigInt): Int =
-    if (this.isValidLong) {
-      if (that.isValidLong) java.lang.Long.compare(this._long, that._long) else -that._bigInteger.signum()
+    if (this.longEncoding) {
+      if (that.longEncoding) java.lang.Long.compare(this._long, that._long) else -that._bigInteger.signum()
     } else {
-      if (that.isValidLong) this.signum else this._bigInteger.compareTo(that._bigInteger)
+      if (that.longEncoding) this.signum else this._bigInteger.compareTo(that._bigInteger)
     }
 
   /** Addition of BigInts
    */
   def +  (that: BigInt): BigInt = {
-    if (this.isValidLong && that.isValidLong) { // fast path
+    if (this.longEncoding && that.longEncoding) { // fast path
       val x = this._long
       val y = that._long
       val z = x + y
-      if ((~(x ^ y) & (x ^ z)) >= 0L) return BigInt(z)
+      if ((~(x ^ y) & (x ^ z)) >= 0L) new BigInt(z)
```

Comment:
aren't you bypassing the ability to save boxed BigInt for small values by using `new` rather than `return` here?

Aslo, I'm confused how not using return works since you don't have else branches, so, isn't this always falling through to the bottom?

If so, the whole if is a waste right? and also isn't this issue replicated below?

Reply:
Bug corrected. Will benchmark the use of the cache or not. (Noting that the original BigInt doesn't create new instances through the `apply` method, and thus doesn't use the cache)